React adapter 16: portals and roots may render fragments

jquense opened the pull request:
This matches what react-test-renderer does as well.

ljharb commented on packages/enzyme-test-suite/test/ReactWrapper-spec.jsx:
```diff
@@ -833,6 +833,17 @@ describeWithDOM('mount', () => {
+        </React.Fragment>
+      );
+
+      expect(wrapper).to.have.lengthOf(2);
```
can we add assertions that the wrapper has a p and a span?

jquense commented:
I removed the two shallow tests, i'm not sure why i added them in the first place. ShallowRenderer doesn't support mounting a Fragment, and a Portal doesn't make any sense, since it's a DOM specific element.

jquense commented on packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js:
```diff
@@ -251,11 +252,13 @@ class ReactSixteenAdapter extends EnzymeAdapter {
           );
-          if (!isStateful) {
+
+          if (!isStateful && typeof Component === 'function') {
```
This is just so we get a clearer error is a user tries to mount a Fragment or Portal, they will see the ShallowTestRenderer  informative error instead of the much less clear "Component is not a function"

ljharb commented:
@jquense I'd actually added them, for test parity. We *should* be able to shallow-render a fragment and a Portal.

jquense commented:
I'm not sure what you want to do in this case, it doesn't make much sense to me to intentionally hobble the mount cases for parities sake in my mind.

The main issue tho is that upstream shallow renderer does not allow it. Same is true for the new context types as well

ljharb commented:
Hmm. I certainly don't want to limit what `mount` can do.

However, I also don't want to limit what `shallow` can do just because the upstream renderer can't handle it. Is there *any* way that a Portal, Fragment, or Context component couldn't just be treated like a `div`, in that in shallow mode it simply renders its children?

jquense commented:
I think we can probably get clever and hack something together but maybe we can split that into a new PR? The goal here was really to support selectors through these types deep in element tree not at the root. at moment the adapter assumes they can only render single children. The tests have sort of accidentally brought to light this inconsistency, but it's one that already exists, not one introduced by this PR
